server/src/main/java/org/elasticsearch/action/search/AbstractSearchAsyncAction.java

```diff
@@ -318,8 +319,11 @@ public final ShardSearchTransportRequest buildShardSearchRequest(SearchShardIter
         AliasFilter filter = aliasFilter.get(shardIt.shardId().getIndex().getUUID());
         assert filter != null;
         float indexBoost = concreteIndexBoosts.getOrDefault(shardIt.shardId().getIndex().getUUID(), DEFAULT_INDEX_BOOST);
-        return new ShardSearchTransportRequest(shardIt.getOriginalIndices(), request, shardIt.shardId(), getNumShards(),
-            filter, indexBoost, timeProvider.getAbsoluteStartMillis(), clusterAlias);
+        int[] indexShards = getIndexShards(shardIt.shardId().getIndex());
```

Comment:
this is quite a complex operation since we call if for every shard and then do consume the entire iterator again. I wonder if we can pre-sort the ShardRoutings in `SearchShardIterator` and then calculate this on the fly and simply call `SearchShardIterator#getIndexShardOrdinal()` to get it?

Reply:
I changed the logic to compute the needed informations only once in `InitialSearchPhase` constructor. I need the complete `GroupShardsIterator` to do so which is why it's not in `SearchShardIterator` but it's the same idea:
https://github.com/elastic/elasticsearch/commit/b90716cfa7eb17b10de0d8398220431fbe856da5